crypto/crypto/src/multi_ed25519.rs

```diff
@@ -0,0 +1,547 @@
+            let count_set_bits = count_set_bits(&temp);
+            if count_set_bits != num_of_sigs || count_set_bits != 0 {
+                return Err(CryptoMaterialError::DeserializationError);
+            }
+            temp
+        } else {
+            let mut temp = BitVec::from_elem(MAX_NUM_OF_KEYS, false);
+            for i in 0..num_of_sigs {
+                temp.set(i, true);
+            }
+            temp
+        };
+
+        let mut signatures: Vec<Ed25519Signature> = vec![];
+        let multi_signature = bytes
+            .chunks(ED25519_SIGNATURE_LENGTH)
+            .filter(|chunk| chunk.len() == ED25519_SIGNATURE_LENGTH)
+            .try_for_each(|sig_bytes| deserialize_and_push(sig_bytes, &mut signatures));
+
+        multi_signature.map(|_| MultiEd25519Signature(signatures, bitvec))
+    }
+}
+
+impl Length for MultiEd25519Signature {
+    fn length(&self) -> usize {
+        self.to_bytes().len()
+    }
+}
+
+impl std::hash::Hash for MultiEd25519Signature {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        let encoded_signature = self.to_bytes();
+        state.write(&encoded_signature);
+    }
+}
+
+// Those are required by the implementation of hash above.
+impl PartialEq for MultiEd25519Signature {
+    fn eq(&self, other: &MultiEd25519Signature) -> bool {
+        self.to_bytes() == other.to_bytes()
+    }
+}
+
+impl Eq for MultiEd25519Signature {}
+
+impl fmt::Display for MultiEd25519Signature {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", hex::encode(&self.to_bytes()[..]))
+    }
+}
+
+impl fmt::Debug for MultiEd25519Signature {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "MultiEd25519Signature({})", self)
+    }
+}
+
+impl ValidKey for MultiEd25519Signature {
+    fn to_bytes(&self) -> Vec<u8> {
+        self.to_bytes()
+    }
+}
+
+//////////////////////
+// Signature Traits //
+//////////////////////
+
+impl Signature for MultiEd25519Signature {
+    type VerifyingKeyMaterial = MultiEd25519PublicKey;
+    type SigningKeyMaterial = MultiEd25519PrivateKey;
+
+    /// Checks that `self` is valid for `message` using `public_key`.
+    fn verify(&self, message: &HashValue, public_key: &MultiEd25519PublicKey) -> Result<()> {
+        self.verify_arbitrary_msg(message.as_ref(), public_key)
+    }
+
+    /// Checks that `self` is valid for an arbitrary &[u8] `message` using `public_key`.
+    /// Outside of this crate, this particular function should only be used for native signature
+    /// verification in Move.
+    fn verify_arbitrary_msg(
+        &self,
+        message: &[u8],
+        public_key: &MultiEd25519PublicKey,
+    ) -> Result<()> {
+        if last_bit_set(&self.1) > public_key.length()
+            || count_set_bits(&self.1) < public_key.1 as usize
+        {
+            return Err(anyhow!("{}", CryptoMaterialError::ValidationError));
```

Comment:
let's try to avoid pulling in anyhow into this crate.

Reply:
Ok, it was already used in `ed25519` and kept it here, but I'll try to come up with something else.

Reply:
thanks! Francois recommendation on thiserror is a solid type for errors throughout this crate.